### PR TITLE
Change pandas materializer to tempfile copy instead of fileio.open

### PR DESCRIPTION
**Pull request by safoinme**
## Describe changes
To allow external custom artifact stores, we change the way pandas materializes loads and saves files by copying the file into a temporary file instead of using fileio.open directly.

## Pre-requisites
Please ensure you have done the following:
- [X] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/component-gallery/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

**Review by htahir1 — COMMENTED** [review]
Are we sure we want to do this just for this case? Why don't we just ask people to create a custom materializer? I think copying is a big overhead tbh and it would be better to have a switch case or something at least

**Comment by safoinme** [comment]
@htahir1 that was exactly the same logic behind using fileio.open instead of copying twice just to run this! while technically even creating a new materialize is sort of additional step that can be last solution since `fileio.open` should be working just fine if implemented right in custom artifact store! 
do you think we just close this one and just suggest to people one of the 2 proper ways of doing this?

**Review by fa9r — COMMENTED** [review]
The changes look fine but I don't understand yet why we need this. Could you give some more background on when exactly the current implementation is not working?

**Comment by safoinme** [comment]
@fa9r "The current implementation requires that any custom artifact store used must have an open function implemented and that the `fileio.open` function can recognize which artifact store to use based on the schema of the given path. If either of these conditions is not met, the materialize function will raise an error because it is unable to open the file."

**Comment by fa9r** [comment]
Sounds to me like "If a custom artifact store is implemented incorrectly it cannot open the file", isn't that intended behaviour? 😄

**Comment by htahir1** [comment]
@fa9r As seen in the comments above, I think we already have a way forward where @safoinme is going to mention this in the docs now and will use his materialzier as an example implementation only :-)

File: src/zenml/materializers/pandas_materializer.py [hunk 1]
```diff
@@ -125,9 +138,20 @@ def save(self, df: Union[pd.DataFrame, pd.Series]) -> None:
...
-            with fileio.open(self.csv_path, mode="wb") as f:
-                df.to_csv(f, index=True)
+            with tempfile.NamedTemporaryFile(
+                mode="wb", suffix=".csv", delete=False
+            ) as f:
+                df.to_csv(f.name, index=True)
+                fileio.copy(f.name, self.csv_path)
+
+        # Close and remove the temporary file
+        f.close()
```
Comment (schustmi) [inline src/zenml/materializers/pandas_materializer.py:156]: I think this isn't needed, the `with ...` context manager should already close the file automatically?

**Pull request closed without merging**